src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,340 @@
-  else
-    Log::Info << "Loading '" << filename << "' as " << stringType << ".  "
-        << std::flush;
-
-  // We can't use the stream if the type is HDF5.
-  bool success;
-  LoadCSV loader;
-
-  if (loadType != FileType::HDF5Binary)
+  else if (opts.TimestampCol() && opts.LabelCol())
   {
-    if (loadType == FileType::CSVASCII)
-      success = loader.LoadNumericCSV(matrix, stream);
-    else
-      success = matrix.load(stream, ToArmaFileType(loadType));
+    matrix.shed_col(0);
+    matrix.shed_col(matrix.n_cols - 1);
+    ReshapeTimeseries(matrix, opts);
   }
-  else
-    success = matrix.load(filename, ToArmaFileType(loadType));
-
-  if (!success)
+  else if (!opts.TimestampCol() && opts.LabelCol())
   {
-    Log::Info << std::endl;
-    Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Loading from '" << filename << "' failed." << std::endl;
-    else
-      Log::Warn << "Loading from '" << filename << "' failed." << std::endl;
-
-    return false;
+    matrix.shed_col(matrix.n_cols - 1);
+    ReshapeTimeseries(matrix, opts);
   }
-  else
-    Log::Info << "Size is " << (transpose ? matrix.n_cols : matrix.n_rows)
-        << " x " << (transpose ? matrix.n_rows : matrix.n_cols) << ".\n";
-
-  // Now transpose the matrix, if necessary.
-  if (transpose)
+  else if (opts.TimestampCol() && !opts.LabelCol())
   {
-    success = inplace_transpose(matrix, fatal);
+    matrix.shed_col(0);
+    ReshapeTimeseries(matrix, opts);
   }
 
-  Timer::Stop("loading_data");
-
-  // Finally, return the success indicator.
+   // Re-Transpose back to faciliate the cases in this function.
+  if (!opts.NoTranspose())
+  {
+    inplace_trans(matrix);
+  }
   return success;
+} 
+
+/*
+ * This function convert timeseries data collected from sensors into specific
+ * Windows that are suitable to training for machine learning.
+ * 
+ * This function is internal and should not be documented or used by users.
+ *
+ * This function only works on non-transposed matrices
+ */
+template<typename eT>
+void ReshapeTimeseries(arma::Mat<eT>& matrix,
+                       CSVOptions& opts)
+{
+  arma::Mat<eT> tmpMat = matrix.as_row();
+  size_t totalWindows = tmpMat.n_elem / opts.WindowSize();
+  //! This operation will give us by default a transposed operation.
+  //! Therefore we need to transpose back.
+  matrix = reshape(tmpMat, opts.WindowSize(), totalWindows);
+  inplace_trans(matrix);
 }
 
-// Load with mappings.  Unfortunately we have to implement this ourselves.
-template<typename eT, typename PolicyType>
+// This is the function that the user is supposed to call.
+// Other functions of this class should be labelled as private.
+template<typename MatType, typename DataOptionsType>
 bool Load(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          DatasetMapper<PolicyType>& info,
-          const bool fatal,
-          const bool transpose)
+          MatType& matrix,
+          DataOptionsType& opts)
```

Comment:
I can think of two ways that you might distinguish this overload from the model type overload:

SFINAE:

```
template<typename MatType, typename DataOptionsType>
bool Load(const std::string& filename,
     MatType& matrix,
     DataOptionsType& opts,
     const std::enable_if_t<arma::is_arma_type<MatType>::value>* = 0);

template<typename ModelType, typename DataOptionsType>
bool Load(const std::string& filename,
    ModelType& m,
    DataOptionsType& opts,
    std::enable_if_t<HasSerialize<ModelType>::value>* = 0);
```

Another way would be to restrict `DataOptionsType` to `ModelOptionsType`.